### PR TITLE
Edit item bitstreams

### PR DESCRIPTION
**Pull request by Atmire-Kristof**
This PR adds an implementation for the bitstreams tab on an item's edit page.

#### UI

Similar to DSpace 6, the page contains a table listing all bitstreams separated by bundle.
You'll notice the columns are about the same, except it contains an "Actions" column at the end instead of the "View" and "Order" columns.
This "Actions" column contains buttons for each bitstream:
- Download - this replaces the "view" column and downloads or displays the bitstream
- Edit - Opens the bitstream's edit page
- Remove - Marks the bitstream for removal
- Undo changes - Reverts any changes made to the bitstream (currently the only change it can revert is the bitstream being marked for removal)

The bundle row also has a button in the action column for uploading a new bitstream to that bundle. This is the same as clicking the upload button on the top-left of the page, but it'll automatically select the relevant bundle in the dropdown.

Instead of the order column, the user is now able to drag bitstreams around within a bundle, using a handle in the front of the bitstream's table tow.
Hovering over a bitstream's row will make the handle appear. Clicking and dragging this handle will move the bitstream to whatever position you drop it on.
If the list of bitstreams within a bundle contains pagination (more than 10 bitstreams), it's possible to drop a bitstream on top of a page number, which will move that bitstream to the top of that page.

All of these changes are kept locally until the user clicks "Save".
Upon clicking "Save", requests (delete and patch) are sent to the REST API, notifications will be displayed and the table is refreshed.

#### Technical details

**Table component structure**
The main component containing the table and submit implementation is `ItemBitstreamsComponent`. This component renders a table header and a `ItemEditBitstreamBundleComponent` for each of the bundles within the item. This component renders the bundle header (name and upload button) and a `PaginatedDragAndDropBitstreamListComponent`.
In turn, this component renders the draggable bitstream rows within a pagination component. The contents of the draggable bitstream row are defined in `ItemEditBitstreamComponent`.

**Paginated drag-and-drop components**
The `PaginatedDragAndDropBitstreamListComponent` extends from a new abstract component: `AbstractPaginatedDragAndDropListComponent`.
This is an abstract component containing general methods and logic to be able to drag and drop objects within a paginated list. In this case Bitstreams, but the implementation could be used for any type of DSpaceObject.

### CODE REVIEW

File: src/app/+item-page/bitstreams/upload/upload-bitstream.component.ts [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+  constructor(protected route: ActivatedRoute,
+              protected router: Router,
+              protected itemService: ItemDataService,
+              protected bundleService: BundleDataService,
+              protected authService: AuthService,
+              protected notificationsService: NotificationsService,
+              protected translate: TranslateService) {
+  }
+
+  ngOnInit(): void {
```
Comment (tdonohue) [inline src/app/+item-page/bitstreams/upload/upload-bitstream.component.ts]: Please add TypeDocs to describe initialization

File: src/app/+item-page/edit-item-page/item-bitstreams/item-edit-bitstream-drag-handle/item-edit-bitstream-drag-handle.component.ts [hunk 2]
```diff
@@ -0,0 +1,21 @@
+import { Component, OnInit, TemplateRef, ViewChild, ViewContainerRef } from '@angular/core';
+
+@Component({
+  selector: 'ds-item-edit-bitstream-drag-handle',
+  styleUrls: ['../item-bitstreams.component.scss'],
+  templateUrl: './item-edit-bitstream-drag-handle.component.html',
+})
+export class ItemEditBitstreamDragHandleComponent implements OnInit {
```
Comment (tdonohue) [inline src/app/+item-page/edit-item-page/item-bitstreams/item-edit-bitstream-drag-handle/item-edit-bitstream-drag-handle.component.ts]: While this is mostly self-explanatory, we should add TypeDocs to describe this component.

File: src/app/core/data/array-move-change-analyzer.service.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
+import { MoveOperation } from 'fast-json-patch/lib/core';
+import { Injectable } from '@angular/core';
+import { moveItemInArray } from '@angular/cdk/drag-drop';
+import { hasValue } from '../../shared/empty.util';
+
+/**
+ * A class to determine move operations between two arrays
+ */
+@Injectable()
+export class ArrayMoveChangeAnalyzer<T> {
```
Comment (tdonohue) [inline src/app/core/data/array-move-change-analyzer.service.ts]: I think this class is missing specs?  Unless I overlooked them.   It'd be important to add specs here to ensure it is generating the correct MoveOperation when diffing Arrays

File: src/app/core/data/object-updates/object-updates.service.ts [hunk 4]
```diff
@@ -138,6 +166,31 @@ export class ObjectUpdatesService {
...
   }
 
+  /**
+   * Method that combines the state's updates with the initial values (when there's no update),
+   * sorted by their custom order to create a FieldUpdates object
+   * @param url The URL of the page for which the FieldUpdates should be requested
+   * @param initialFields The initial values of the fields
+   * @param page The page to retrieve
+   */
+  getFieldUpdatesByCustomOrder(url: string, initialFields: Identifiable[], page = 0): Observable<FieldUpdates> {
```
Comment (tdonohue) [inline src/app/core/data/object-updates/object-updates.service.ts]: I think this method is missing specs?

File: src/app/core/data/object-updates/object-updates.service.ts [hunk 5]
```diff
@@ -330,4 +405,19 @@ export class ObjectUpdatesService {
   getLastModified(url: string): Observable<Date> {
     return this.getObjectEntry(url).pipe(map((entry: ObjectUpdatesEntry) => entry.lastModified));
   }
+
+  /**
+   * Get move operations based on the custom order
+   * @param url The page's url
+   */
+  getMoveOperations(url: string): Observable<MoveOperation[]> {
```
Comment (tdonohue) [inline src/app/core/data/object-updates/object-updates.service.ts]: Again, I think this one is missing from the specs.  The Spec changes for this service were a lot smaller than the service changes.

File: src/app/core/shared/item-bitstreams-utils.ts [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+import { RemoteData } from '../data/remote-data';
+import { PaginatedList } from '../data/paginated-list';
+import { Bitstream } from './bitstream.model';
+import { map } from 'rxjs/operators';
+import { hasValue, hasValueOperator } from '../../shared/empty.util';
+
+/**
+ * Operator for turning the current page of bitstreams into an array
+ */
+export const toBitstreamsArray = () =>
```
Comment (tdonohue) [inline src/app/core/shared/item-bitstreams-utils.ts]: Similar to the existing `metadata.utils.ts`, I think this file should be named `item-bitstreams.utils.ts`. 

Also, I think this `toBitstreamsArray` is missing specs.

File: src/app/shared/form/builder/ds-dynamic-form-ui/models/custom-switch/custom-switch.component.ts [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+import { Component, EventEmitter, Input, Output } from '@angular/core';
+import { FormGroup } from '@angular/forms';
+import { DynamicCustomSwitchModel } from './custom-switch.model';
+
+@Component({
+  selector: 'ds-custom-switch',
+  styleUrls: ['./custom-switch.component.scss'],
+  templateUrl: './custom-switch.component.html',
+})
+export class CustomSwitchComponent extends DynamicNGBootstrapCheckboxComponent {
```
Comment (tdonohue) [inline src/app/shared/form/builder/ds-dynamic-form-ui/models/custom-switch/custom-switch.component.ts]: Please add TypeDocs to describe this new component.  Also, I'm not seeing any specs related to this component.

File: src/app/shared/form/builder/ds-dynamic-form-ui/models/custom-switch/custom-switch.model.ts [hunk 8]
```diff
@@ -0,0 +1,16 @@
+import {
+  DynamicCheckboxModel,
+  DynamicCheckboxModelConfig,
+  DynamicFormControlLayout,
+  serializable
+} from '@ng-dynamic-forms/core';
+
+export const DYNAMIC_FORM_CONTROL_TYPE_CUSTOM_SWITCH = 'CUSTOM_SWITCH';
+
+export class DynamicCustomSwitchModel extends DynamicCheckboxModel {
```
Comment (tdonohue) [inline src/app/shared/form/builder/ds-dynamic-form-ui/models/custom-switch/custom-switch.model.ts]: Please add TypeDocs to this new model as well

**Comment by tdonohue** [comment]
@Atmire-Kristof : Could you please rebase this PR.  It's got some minor code conflicts.  Also, Travis should now approve it (as the prior failure was not a fault of the PR itself & I've updated Travis settings so that this should now succeed).

**Comment by benbosman** [comment]
@Atmire-Kristof I've tested this functionality and noticed:
* I I try to upload a file to a **new** bundle (so clicking the main upload button, typing a new bundle name, and uploading a file), I get an exception because a POST to /placeholder is attempted
* Similar to Tim's error, uploading a large file fails, but this failure occurs in REST, so not a problem with this PR

**Comment by benbosman** [comment]
@Atmire-Kristof I'm still encountering problems when uploading a file to a new bundle, but this time a different error

When uploading a bitstream to an existing bundle, it works fine
When uploading a bitstream to a new bundle, I get CORS errors:
> Access to XMLHttpRequest at 'https://dspace7-demo.atmire.com/server/api/core/bundles/9b8fc0a1-98da-4ba7-9d69-d607bc088f0c/bitstreams' from origin 'http://localhost:3000' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: The value of the 'Access-Control-Allow-Origin' header in the response must not be the wildcard '*' when the request's credentials mode is 'include'. The credentials mode of requests initiated by the XMLHttpRequest is controlled by the withCredentials attribute.

When ignoring the CORS headers, the upload succeeds, but the UI doesn't update after completing the update

**Comment by Atmire-Kristof** [comment]
I've done some further testing on this PR with our internal REST API containing the necessary changes to CORS, but I've discovered another issue. Even though the upload succeeds and the bitstream is now added to the new bundle, the UI still doesn't display any progress and doesn't redirect the user to the next page.
For some reason the uploader's `onCompleteItem` event is not triggered after the upload. I'll have to fix this issue first, before we can merge this PR.

**Comment by tdonohue** [comment]
Merging as this is now at +2. I've logged the bug that @benbosman noted above as a separate issue: https://github.com/DSpace/dspace-angular/issues/641

**Comment by tdonohue** [comment]
@Atmire-Kristof : Apologies, I seem to have not seen your comment here before pressing merge on the PR.  If you could create a followup PR to fix [the issue you noted](https://github.com/DSpace/dspace-angular/pull/577#issuecomment-607928607), we can prioritize that for quick review. Thanks!

**Pull request merged**